feat: tracer: upgrade elastic search transport for pubsub traces

cortze opened the pull request:
## Related Issues
The current lotus trace wrapper in [`master`](https://github.com/filecoin-project/lotus/blob/64b9b532c0f8cb55d84be79b6d127c3213deadf5/node/modules/lp2p/pubsub.go#L508) has a limited number of traces pushed to the Elastic Search (ES) remote instance. The daemon only pushes traces related to:
- Join/leave GossipSub topics
- Graft/Prune peers per topic
- GossipSub PeerScores 

Furthermore, the current code also forces the user to set up a [`RemoteTrace`](https://github.com/filecoin-project/lotus/blob/6c0a2feade9ac2c5d9d35a31839a606ff94c0179/node/modules/lp2p/pubsub.go#L4272) in the `config-file` to track the published messages of the desired topics.

Related to the proposed gossipsub measurement study [protocol/network-measurements#17](https://github.com/protocol/network-measurements/issues/17), to track broadcasting latencies of gossipsub messages, the `PRC_Recv` events are needed. Thus, this PR enables the remote submission of RPC-related traces. As RPC calls already include all the `mesh-control` messages, these ones are disabled in pro of reducing the overall overhead.

## Proposed Changes
Submitting all the RPC traces to the remote ES instance through individual HTTP requests adds significant overhead to the lotus daemon, making it lose synchronization with the head of the chain. Thus, this PR upgrades the Elastic Search transport to support trace batching and a faster HTTP transport in the ES client.

The PR applies some parameters that would work fine and are currently being tested. Feel free to suggest better approaches or parameters like the `flushing time` or `buffer limit` for the batching system.

## Checklist

Before you mark the PR ready for review, please make sure that:

- [x] Commits have a clear commit message.
- [x] PR title is in the form of `<PR type>: <area>: <change being made>`
- [x] New features have usage guidelines and / or documentation updates in
  - [x] [Lotus Documentation](https://lotus.filecoin.io)
  - [x] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [x] Tests exist for new functionality or change in behavior
- [x] CI is green

cortze commented:
# Updates 
The suggested changes have been running for 5 consecutive days on a locally monitored node. To bring more information about how the remote traces submission impacts the overall performance of the Lotus daemon, these are some plots I generated when comparing the resources used the `lotus@v1.19.0` after #7398 with the proposed changes.

NOTE: 
- All the graphs include two different distributions delimited by the `red` vertical line on `09/03/2023 at 16pm`. 
- The distribution on the left corresponds to the data gathered from the `Prometheus node exporter` running on the same machine as the Lotus daemon on its `v1.19.0` after #7398 , used to sync the filecoin chain from the latest snapshot available. 
- The distribution on the right corresponds to the compiled lotus daemon with the proposed changes

## Highlights
- The lotus daemon maintains the same level of resource utilization with the major exception of the network traffic (both incoming and outcoming).
- The remote trace submission to an ES instance is the one that generates the extra measured traffic. 
- In both cases, the lotus daemon follows the head of the chain without any apparent problem (no WARN-ERROR logs & `lotus sync` always reporting `0` blocks missing)

## Resource profiling 
- Disk Usage: The aggregated disk usage of the lotus client remains constant at around 4GB/h

| Disk space increment |  Increment per hour: |  
|---|---|
| ![image](https://user-images.githubusercontent.com/45786396/224959591-ca18b85c-aa79-4845-af85-cae40536cbcd.png)  | ![image](https://user-images.githubusercontent.com/45786396/224959611-a5f6f0ef-c89a-460e-b5b9-590f5a514448.png) |

- Memory usage: the allocated memory by the process follows the exact same increment over time.

<img src="https://user-images.githubusercontent.com/45786396/224960928-19a23264-6517-4167-a2c5-da86a201bb1d.png" height="auto" width="400">

- CPU usage: There are no significant differences in CPU usage when tracking the PubSub Traces (~2.5% extra at some point of `12/03` ). Note that the first spike of CPU usage is due to the synchronization of the chain.
 
<img src="https://user-images.githubusercontent.com/45786396/224962456-87cff82a-4cd1-44ac-be8e-1ace403fd868.png " height="auto" width="400">

- Net-In: x1.5-x2 traffic increment (~150MB/h) when enabling the remote trace submission to the ES instance. Note the initial spike originated from the lotus node syncing the chain and that the same network increment is measured as Net-Out at the ES machine:

| Lotus Net-In | ES Net-Out |
|---|---| 
| ![image](https://user-images.githubusercontent.com/45786396/224962836-df1d3f39-2cc7-4330-a119-ad1dafc517aa.png) | ![image](https://user-images.githubusercontent.com/45786396/224963574-c7449c9f-ee95-4c0b-801e-c55b2134dfbf.png) |


- Net-Out: x2 increase in sent bytes (350MB/h) after enabling the remote gossip traces submission to ES, probably generated from formatting GossipSub traces in json packages and submitting them to a remote Elastic-Search instance. 

| Lotus Net-Out | ES Net-In |
|---|---| 
| ![image](https://user-images.githubusercontent.com/45786396/224962799-4f401315-e854-4b10-a6c6-f250ab5ce201.png) | ![image](https://user-images.githubusercontent.com/45786396/224963742-6524428d-c514-44d3-9e14-8d710bbd2559.png) |

cortze commented:
Rebased PR to `v1.22.0-rc4` as it is the latest release

snadrus commented on node/modules/lp2p/pubsub.go:
```diff
@@ -486,71 +498,169 @@ func (trw *tracerWrapper) Trace(evt *pubsub_pb.TraceEvent) {
+			}
+		}
+		// check if we have any of iwant msgs (it doesn't classify per topic - just msg.ID)
+		var validIwant bool = false
```
Simpler:
validIwant := len(iwant) > 0

cortze replied:
got it! thanks for the simplification

cortze commented on node/modules/tracer/elasticsearch_transport.go:
```diff
@@ -30,10 +34,10 @@ func NewElasticSearchTransport(connectionString string, elasticsearchIndex strin
 		},
 		Username: username,
 		Password: password,
+		Transport: &FastHttpTransport{},
```
Totally fine with the change. Let me address both comments and test it on my test env :)

cortze commented on node/modules/lp2p/pubsub.go:
```diff
@@ -486,71 +498,169 @@ func (trw *tracerWrapper) Trace(evt *pubsub_pb.TraceEvent) {
+			}
+		}
+		// check if we have any of iwant msgs (it doesn't classify per topic - just msg.ID)
+		var validIwant bool = false
```
got it! thanks for the simplification

cortze commented:
Hey, @snadrus ! Thank you so much for the feedback! I just addressed your comments bringing back the `net/http` transport. 
Let me know if anything is missing from my side :)
